File: 백준/Silver/9012. 괄호/괄호.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const T = +input[0];

input.shift();

for (let i = 0; i < T; i++) {
  let str = input[i].replace("\r", "");
  let result = "YES";
  let count = [];

  for (let j = 0; j < str.length; j++) {
    if (str[j] === "(") {
      count.push(1);
    } else if (!count.pop()) {
      result = "NO";
      break;
    }
  }
  if (count.length !== 0) {
    result = "NO";
  }
  console.log(result);
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringBuilder sb = new StringBuilder();

		int T = Integer.parseInt(br.readLine());

		for (int tc = 0; tc < T; tc++) {

			String[] s = br.readLine().split("");
			int count = 0;

			for (int i = 0; i < s.length; i++) {

				if (s[i].equals("(")) {
					count++;
				} else {
					count--;
				}
							
				if(count < 0) {
					sb.append("NO").append("\n");
					break;
				}

			}
			
			if(count > 0) {
				sb.append("NO").append("\n");
			}else if(count == 0) {
				sb.append("YES").append("\n");
			}

		}

		bw.write(sb.toString());
		bw.flush();
		bw.close();
		br.close();
	}

}